guardian/managers.py

```diff
@@ -67,14 +81,43 @@ def bulk_assign_perm(self, perm, user_or_group, queryset):
         assigned_perms = []
         for instance in queryset:
             if not checker.has_perm(permission.codename, instance):
-                kwargs = {'permission': permission, self.user_or_group_field: user_or_group}
-                if self.is_generic():
-                    kwargs['content_type'] = ctype
-                    kwargs['object_pk'] = instance.pk
-                else:
-                    kwargs['content_object'] = instance
+                kwargs = self._perm_kwargs(permission, user_or_group, instance, ctype, user_group_object)
                 assigned_perms.append(self.model(**kwargs))
-        self.model.objects.bulk_create(assigned_perms)
+        self.bulk_create(assigned_perms)
+
+        return assigned_perms
+
+    def bulk_assign_perm_from_user_group_objects(self, perm, user_group_objects):
+        """
+        Bulk assigns permissions with given ``perm`` for the user and content_object in UserGroupObjects.
+        The content_objects must be of the same type.
+        """
+        if len(user_group_objects) == 0:
+            return []
+        ctype = get_content_type(user_group_objects[0].content_object)
+
+        if not isinstance(perm, Permission):
+            permission = Permission.objects.get(content_type=ctype, codename=perm)
+        else:
+            permission = perm
+
+        by_user = {o.user: [] for o in user_group_objects}
+        for o in user_group_objects:
+            by_user[o.user].append(o)
+
+        assigned_perms = []
+        for user in by_user:
+            objects = by_user[user]
+
+            checker = ObjectPermissionChecker(user)
+            checker.prefetch_perms([o.content_object for o in objects])
```

Comment:
I'll add this check

if isinstance(origins, QuerySet):
            origins = origins.select_related('content_object')